pkg/events/default.go

```diff
@@ -14,24 +14,50 @@
 
 package events
 
-import "context"
+import (
+	"context"
+	"sync"
+)
 
-// DefaultPubSub is the default event pubsub.
-var DefaultPubSub = NewPubSub(DefaultBufferSize)
+var defaultPubSubMu = &sync.Mutex{}
+
+// defaultPubSub is the default event pubsub.
+var defaultPubSub = NewPubSub(DefaultBufferSize)
+
+// SetDefaultPubSub sets pubsub used by the package to ps and returns a function,
+// which can be used to undo the action.
+func SetDefaultPubSub(ps PubSub) func() {
+	defaultPubSubMu.Lock()
```

Comment:
If production code calls `SetDefaultPubSub` without calling the undo func we'll likely have a deadlock. That should either not happen or be documented properly.

Reply:
Also, why is this mutex only used for write and not for read? Are we trying to prevent concurrent access to `defaultPubSub` or are we trying to limit the number of times `SetDefaultPubSub` can be called?

Reply:
We're trying to prevent `SetDefaultPubSub` from calling multiple times in production

Reply:
Note, that in some use cases(e.g. testing) we, in fact, want to call `SetDefaultPubSub` multiple times and there we can use this to ensure that the last assignment is "done", and so it's safe to set the `defaultPubSub` to something else

Reply:
It's not the responsibility of the events package to prevent implementations from replacing the default pubsub. Of course it's bad practice to do so after it's being used, but this is not the problem of the events package.

Even if it were the event package's responsibility, a deadlock is definitely the wrong way to tell implementations that they're doing something wrong.